src/pyhf/optimize/mixins.py

```diff
@@ -0,0 +1,140 @@
+        self.verbose = kwargs.pop('verbose', False)
+        self._minimizer = None
+
+        if kwargs:
+            raise KeyError(
+                f"""Unexpected keyword argument(s): '{"', '".join(kwargs.keys())}'"""
+            )
+
+    def _internal_minimize(
+        self,
+        func,
+        init,
+        method='SLSQP',
+        jac=None,
+        bounds=None,
+        fixed_vals=None,
+        options={},
+    ):
+        result = self._minimize(
+            func,
+            init,
+            method=method,
+            bounds=bounds,
+            fixed_vals=fixed_vals,
+            options=options,
+            jac=jac,
+        )
+
+        try:
+            assert result.success
+        except AssertionError:
+            log.error(result)
+            raise
+        return result
+
+    def minimize(
+        self,
+        objective,
+        data,
+        pdf,
+        init_pars,
+        par_bounds,
+        fixed_vals=None,
+        return_fitted_val=False,
+        do_grad=False,
+        do_stitch=False,
+        method='SLSQP',
+        **kwargs,
+    ):
+        """
+        Find Function Parameters that minimize the Objective.
+
+        Args:
+            objective: objective function
+            data: observed data
+            pdf: model
+            init_pars: initial parameters
+            par_bounds: parameter boundaries
+            fixed_vals: fixed parameter values
+            return_fitted_val: return bestfit value of the objective
+            do_grad (`bool`): enable autodifferentiation mode. Default is off.
+            do_stitch (`bool`): enable splicing/stitching fixed parameter.
+            method: minimization routine
+            kwargs: other options to pass through to underlying minimizer
+
+        Returns:
+            bestfit parameters
+
+        """
+        tensorlib, _ = get_backend()
+        tv, fixed_values_tensor, func_and_grad, init, bounds = shim(
+            objective,
+            data,
+            pdf,
+            init_pars,
+            par_bounds,
+            fixed_vals,
+            do_grad=do_grad,
+            do_stitch=do_stitch,
+        )
+        if do_grad:
+            func = lambda pars: func_and_grad(pars)[0]
+            jac = lambda pars: func_and_grad(pars)[1]
+        else:
+            func = func_and_grad
+            jac = None
+
+        self._setup_minimizer(func, data, pdf, init_pars, par_bounds, fixed_vals)
+
+        minimizer_kwargs = dict(method=method, bounds=bounds, options=kwargs, jac=jac)
+        if not do_stitch:
+            minimizer_kwargs.update(dict(fixed_vals=fixed_vals))
+        result = self._internal_minimize(func, init, **minimizer_kwargs)
+
+        nonfixed_vals = tensorlib.astensor(result.x)
+        fitted_val = result.fun
+        # stitch things back up if needed
+        if do_stitch:
+            fitted_pars = tv.stitch([fixed_values_tensor, nonfixed_vals])
+        else:
+            fitted_pars = nonfixed_vals
+
```

Comment:
this should be wrapped into a `postprocess` function

Reply:
I'm not sure we truly need a separate post-process function. If it's a matter of just making the function less complicated, we can add a `_postprocess(result, do_stitch=False, return_fitted_val=False)` to split the `def OptimizerMixin.minimize` up.

Reply:
see above,.. if the shim returns closures it might be quite nice 

```
wrapped_objective, minimizer_kwargs, postprocess = shim()
result = self._minimize(wrapped_objective, **minimizer_kwargs)
result = postprocess(result)
```

Reply:
done